server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1074,35 +1083,29 @@ public void onFailure(String source, Exception e) {
-                                    .findFirst();
-                                // If we can't find the snapshot by the given name in the repository at all or if the snapshot we find in
-                                // the repository is not the one we expected to find when waiting for a finishing snapshot we fail.
-                                if (matchedEntry.isPresent()) {
-                                    deleteCompletedSnapshot(
-                                        new Snapshot(repositoryName, matchedEntry.get()), repositoryData.getGenId(), Priority.NORMAL, l);
-                                } else {
-                                    l.onFailure(new SnapshotMissingException(repositoryName, snapshotName));
-                                }
-                            }, l::onFailure))));
+                    resolveAndDeleteSnapshots(snapshotNames, listener, repositoryName);
                     return;
                 }
                 logger.trace("adding snapshot completion listener to wait for deleted snapshot to finish");
                 addListener(runningSnapshot, ActionListener.wrap(
                     result -> {
                         logger.debug("deleted snapshot completed - deleting files");
-                        deleteCompletedSnapshot(
-                            new Snapshot(repositoryName, result.v2().snapshotId()), result.v1().getGenId(), Priority.IMMEDIATE, listener);
+                        final RepositoryData newRepoData = result.v1();
+                        final List<SnapshotId> foundSnapshots = matchingSnapshotIds(newRepoData, snapshotNames, repositoryName);
+                        assert foundSnapshots.contains(runningSnapshot.getSnapshotId()) : "Expected new RepositoryData to contain ["
+                                + runningSnapshot + "] but it only contained snapshots " + newRepoData.getSnapshotIds();
+                        deleteCompletedSnapshots(foundSnapshots, repositoryName, newRepoData.getGenId(), Priority.IMMEDIATE, listener);
                     },
                     e -> {
                         if (abortedDuringInit) {
                             logger.info("Successfully aborted snapshot [{}]", runningSnapshot);
-                            listener.onResponse(null);
+                            final Collection<String> remainingPatterns = new ArrayList<>(snapshotNames);
+                            remainingPatterns.remove(runningSnapshot.getSnapshotId().getName());
+                            if (remainingPatterns.isEmpty()) {
+                                listener.onResponse(null);
+                                return;
+                            }
+                            resolveAndDeleteSnapshots(remainingPatterns, listener, repositoryName);
                         } else {
                             if (ExceptionsHelper.unwrap(e, NotMasterException.class, FailedToCommitClusterStateException.class)
```

Comment:
You could make the argument that if the snapshot you aborted was matched by a pattern instead of its exact name, then the below handling could just quietly ignore whatever exception made the snapshot never materialize. I opted for retaining failure here since it's only possible to get here by running into some IO issue with the repository and I did not want to fire off another delete request for unmatched patterns after running into an IO issue with the repo.